File: src/Constraints.cpp
/*
 * Copyright 2020 Maximilian Azendorf
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "Constraints.h"

#include "Util.h"
#include "UnionFind.h"

vector<vector<int>> Constraints::get_mandatory_critical_sets(vector<Constraint> const& constraints)
{
    vector<vector<int>> res;
    map<int, vector<int>> constGroups;

    for(Constraint const& constraint : constraints)
    {
        if(constraint.type() != ChooserIsInChoice) continue;
        constGroups[constraint.left()].push_back(constraint.right());
    }

    for(auto const& constGroup : constGroups)
    {
        res.push_back(constGroup.second);
    }

    return res;
}

vector<Constraint> Constraints::expand_dependent_constraints(vector<Constraint> const& constraints, int choiceCount)
{
    vector<Constraint> res;

    for(Constraint constraint : constraints)
    {
        res.push_back(constraint);
    }


    for (auto const& dependance : get_dependent_choices(constraints, choiceCount))
    {
        res.push_back(Constraint(ChoicesAreNotInSameSlot, dependance.first, dependance.second));
    }

    res.erase(std::unique(res.begin(), res.end()), res.end());
    return res;
}

vector<pair<int, int>> Constraints::get_dependent_choices(vector<Constraint> const& constraints, int choiceCount)
{
    vector<pair<int, int>> result;
    UnionFind<int> choiceGroups(choiceCount);

    for(Constraint const& constraint : constraints)
    {
        if(constraint.type() != ChoosersOfChoicesRelation && constraint.extra() != Eq) continue;
        switch(constraint.extra())
        {
            case Eq:
                choiceGroups.join(constraint.left(), constraint.right());
                break;
            case Subset:
            case Superset:
                result.push_back(std::make_pair(constraint.left(), constraint.right()));
                break;
            default:
                throw std::logic_error("Unsupported operation " + str(constraint.extra()) + ".");
        }
    }

    for(auto const& groupList : {choiceGroups.groups(), get_mandatory_critical_sets(constraints)})
    {
        for (auto const& group : choiceGroups.groups())
        {
            for (int i = 0; i < group.size(); i++)
            {
                for (int j = i + 1; j < group.size(); j++)
                {
                    result.push_back(std::make_pair(group[i], group[j]));
                }
            }
        }
    }

    return result;
}

vector<Constraint>
Constraints::reduce_and_optimize(vector<Constraint> const& constraints, int choiceCount, bool& isInfeasible)
{
    isInfeasible = false;
    vector<Constraint> res;

    for(Constraint c : constraints)
    {
        ConstraintType newType = c.type();
        bool switchSides = true;
        bool add = true;

        switch(c.type())
        {
            case SlotContainsChoice: newType = ChoiceIsInSlot; break;
            case SlotNotContainsChoice: newType = ChoiceIsNotInSlot; break;
            case ChoiceContainsChooser: newType = ChooserIsInChoice; break;
            case ChoiceNotContainsChooser: newType = ChooserIsNotInChoice; break;

            case SlotsHaveSameChoices:
            {
                // This constraint is always either a tautology or a contradiction.
                //
                add = false;
                if(c.left() != c.right())
                {
                    isInfeasible = true;
                }
                break;
            }

            default:
            {
                switchSides = false;
                break;
            }
        }

        if(add)
        {
            res.push_back(Constraint(newType, switchSides ? c.right() : c.left(), switchSides ? c.left() : c.right(), c.extra()));
        }
    }

    return expand_dependent_constraints(res, choiceCount);
}

File: src/constraints.rs
use crate::Constraint;
use crate::ConstraintType;
use crate::UnionFind;

pub(crate) fn get_dependent_choices(
    constraints: &[Constraint],
    choice_count: usize,
) -> Vec<Vec<usize>> {
    let mut groups = UnionFind::<usize>::new(choice_count);
    for constraint in constraints {
        if constraint.kind == ConstraintType::ChoicesHaveSameChoosers {
            groups.join(
                constraint.left,
                usize::try_from(constraint.right).expect("choice index must be non-negative"),
            );
        }
    }
    groups.groups()
}

pub(crate) fn reduce_and_optimize(
    constraints: &[Constraint],
    choice_count: usize,
) -> (Vec<Constraint>, bool) {
    let mut is_infeasible = false;
    let mut reduced = Vec::new();

    for constraint in constraints {
        let mut new_kind = constraint.kind;
        let mut switch_sides = true;
        let mut add = true;

        match constraint.kind {
            ConstraintType::SlotContainsChoice => new_kind = ConstraintType::ChoiceIsInSlot,
            ConstraintType::SlotNotContainsChoice => new_kind = ConstraintType::ChoiceIsNotInSlot,
            ConstraintType::ChoiceContainsChooser => new_kind = ConstraintType::ChooserIsInChoice,
            ConstraintType::ChoiceNotContainsChooser => {
                new_kind = ConstraintType::ChooserIsNotInChoice;
            }
            ConstraintType::SlotsHaveSameChoices => {
                add = false;
                if i32::try_from(constraint.left).expect("constraint index must fit in i32")
                    != constraint.right
                {
                    is_infeasible = true;
                }
            }
            _ => {
                switch_sides = false;
            }
        }

        if add {
            reduced.push(Constraint::new(
                new_kind,
                if switch_sides {
                    usize::try_from(constraint.right)
                        .expect("constraint index must be non-negative")
                } else {
                    constraint.left
                },
                if switch_sides {
                    i32::try_from(constraint.left).expect("constraint index must fit in i32")
                } else {
                    constraint.right
                },
                constraint.extra,
            ));
        }
    }

    (
        expand_dependent_constraints(&reduced, choice_count),
        is_infeasible,
    )
}

fn expand_dependent_constraints(
    constraints: &[Constraint],
    choice_count: usize,
) -> Vec<Constraint> {
    let mut result = Vec::new();

    let dependent_choices = get_dependent_choices(constraints, choice_count);
    let mandatory_critical_sets = get_mandatory_critical_sets(constraints);

    for group_list in [&dependent_choices, &mandatory_critical_sets] {
        for group in group_list {
            for i in 0..group.len() {
                for j in (i + 1)..group.len() {
                    result.push(Constraint::new(
                        ConstraintType::ChoicesAreNotInSameSlot,
                        group[i],
                        i32::try_from(group[j]).expect("choice index must fit in i32"),
                        0,
                    ));
                }
            }
        }
    }

    for constraint in constraints {
        if !matches!(
            constraint.kind,
            ConstraintType::ChooserIsInChoice | ConstraintType::ChooserIsNotInChoice
        ) {
            continue;
        }

        let mut group = Vec::new();
        for dep_group in &dependent_choices {
            if dep_group.contains(
                &usize::try_from(constraint.right).expect("choice index must be non-negative"),
            ) {
                group.clone_from(dep_group);
                break;
            }
        }

        if group.is_empty() {
            continue;
        }

        for choice in group {
            if i32::try_from(choice).expect("choice index must fit in i32") == constraint.right {
                continue;
            }
            result.push(Constraint::new(
                constraint.kind,
                constraint.left,
                i32::try_from(choice).expect("choice index must fit in i32"),
                0,
            ));
        }
    }

    result.extend(constraints.iter().copied());
    result.sort_by_key(|constraint| {
        (
            constraint.kind as i32,
            constraint.left,
            constraint.right,
            constraint.extra,
        )
    });
    result.dedup();
    result
}

fn get_mandatory_critical_sets(constraints: &[Constraint]) -> Vec<Vec<usize>> {
    let mut groups = std::collections::BTreeMap::<usize, Vec<usize>>::new();
    for constraint in constraints {
        if constraint.kind == ConstraintType::ChooserIsInChoice {
            groups.entry(constraint.left).or_default().push(
                usize::try_from(constraint.right).expect("choice index must be non-negative"),
            );
        }
    }
    groups.into_values().collect()
}
